File: lessons/07-oop/oop.pl
% Lesson 7: Object-Oriented Programming in Prolog
%
% Prolog is a logic programming language - fundamentally different from OOP!
% However, we can model object-oriented concepts using:
% - Facts (to represent objects and their properties)
% - Rules (to represent methods/behaviors)
% - Predicates (to query object state)
%
% This demonstrates how different paradigms can model similar concepts.

:- initialization(main).

% ====================
% 1. Objects as Facts (Basic "Class")
% ====================

% person(Name, Age) - represents a Person object
person(alice, 30).
person(bob, 25).
person(charlie, 35).

% "Method" - introduce yourself
introduce(Person) :-
    person(Person, Age),
    format('   Hi, I\'m ~w, ~w years old~n', [Person, Age]).

% "Method" - check if adult
is_adult(Person) :-
    person(Person, Age),
    Age >= 18.

% ====================
% 2. "Inheritance" with Type Hierarchy
% ====================

% Base "class" - animal
animal(buddy, dog, canine).
animal(whiskers, cat, feline).
animal(max, dog, canine).
animal(tweety, bird, avian).

% "Subclass" properties
dog_breed(buddy, 'Golden Retriever').
dog_breed(max, 'German Shepherd').

cat_indoor(whiskers, true).

% "Method" - polymorphic speak
speak(Name, Sound) :-
    animal(Name, dog, _),
    format('   ~w says Woof!~n', [Name]).

speak(Name, Sound) :-
    animal(Name, cat, _),
    format('   ~w says Meow!~n', [Name]).

speak(Name, Sound) :-
    animal(Name, bird, _),
    format('   ~w says Tweet!~n', [Name]).

speak(Name, Sound) :-
    animal(Name, Type, _),
    \+ member(Type, [dog, cat, bird]),
    format('   ~w says Some generic animal sound~n', [Name]).

% "Method" - sleep (default implementation for all animals)
sleep(Name) :-
    animal(Name, _, _),
    format('   ~w is sleeping... Zzz~n', [Name]).

% Dog-specific "method"
fetch(Name) :-
    animal(Name, dog, _),
    format('   ~w is fetching the ball!~n', [Name]).

% Cat-specific "method"
scratch(Name) :-
    animal(Name, cat, _),
    format('   ~w is scratching the furniture!~n', [Name]).

% ====================
% 3. Shapes with Computed Properties
% ====================

% shape(ID, Type, Properties)
shape(circle1, circle, [radius-5]).
shape(rect1, rectangle, [width-4, height-6]).
shape(tri1, triangle, [side_a-3, side_b-4, side_c-5]).

% "Method" - calculate area (polymorphic)
area(ShapeID, Area) :-
    shape(ShapeID, circle, Props),
    member(radius-R, Props),
    Area is pi * R * R.

area(ShapeID, Area) :-
    shape(ShapeID, rectangle, Props),
    member(width-W, Props),
    member(height-H, Props),
    Area is W * H.

area(ShapeID, Area) :-
    shape(ShapeID, triangle, Props),
    member(side_a-A, Props),
    member(side_b-B, Props),
    member(side_c-C, Props),
    S is (A + B + C) / 2,
    Area is sqrt(S * (S - A) * (S - B) * (S - C)).

% "Method" - calculate perimeter
perimeter(ShapeID, Perim) :-
    shape(ShapeID, circle, Props),
    member(radius-R, Props),
    Perim is 2 * pi * R.

perimeter(ShapeID, Perim) :-
    shape(ShapeID, rectangle, Props),
    member(width-W, Props),
    member(height-H, Props),
    Perim is 2 * (W + H).

perimeter(ShapeID, Perim) :-
    shape(ShapeID, triangle, Props),
    member(side_a-A, Props),
    member(side_b-B, Props),
    member(side_c-C, Props),
    Perim is A + B + C.

% "Method" - describe shape
describe_shape(ShapeID) :-
    shape(ShapeID, Type, _),
    area(ShapeID, A),
    perimeter(ShapeID, P),
    format('   ~w: area=~2f, perimeter=~2f~n', [Type, A, P]).

% ====================
% 4. Encapsulation with Dynamic Predicates
% ====================

:- dynamic account/3.  % account(ID, AccountNumber, Balance)
:- dynamic transaction/2.  % transaction(ID, List)

% Initialize account counter
:- dynamic account_counter/1.
account_counter(0).

% "Constructor" - create new account
new_account(AccountNumber, InitialBalance, ID) :-
    retract(account_counter(Counter)),
    ID is Counter + 1,
    assert(account_counter(ID)),
    assert(account(ID, AccountNumber, InitialBalance)),
    assert(transaction(ID, [])).

% "Method" - deposit
deposit(ID, Amount) :-
    Amount > 0,
    retract(account(ID, AccNum, Balance)),
    NewBalance is Balance + Amount,
    assert(account(ID, AccNum, NewBalance)),
    retract(transaction(ID, Trans)),
    format(atom(Msg), 'Deposit: +$~2f', [Amount]),
    append(Trans, [Msg], NewTrans),
    assert(transaction(ID, NewTrans)).

% "Method" - withdraw
withdraw(ID, Amount) :-
    account(ID, AccNum, Balance),
    Amount > 0,
    Amount =< Balance,
    retract(account(ID, AccNum, Balance)),
    NewBalance is Balance - Amount,
    assert(account(ID, AccNum, NewBalance)),
    retract(transaction(ID, Trans)),
    format(atom(Msg), 'Withdrawal: -$~2f', [Amount]),
    append(Trans, [Msg], NewTrans),
    assert(transaction(ID, NewTrans)).

% "Method" - get balance
get_balance(ID, Balance) :-
    account(ID, _, Balance).

% "Method" - get transactions
get_transactions(ID, Trans) :-
    transaction(ID, Trans).

% ====================
% 5. Temperature Conversion (Functional-style)
% ====================

temperature(celsius, C, C).
temperature(fahrenheit, F, C) :- C is (F - 32) * 5 / 9.
temperature(kelvin, K, C) :- C is K - 273.15.

to_fahrenheit(C, F) :- F is C * 9 / 5 + 32.
to_kelvin(C, K) :- K is C + 273.15.
is_freezing(C) :- C =< 0.

% ====================
% 6. Composition
% ====================

% engine(EngineID, Horsepower, Running)
:- dynamic engine/3.
:- dynamic engine_counter/1.
engine_counter(0).

% Create new engine
new_engine(Horsepower, ID) :-
    retract(engine_counter(Counter)),
    ID is Counter + 1,
    assert(engine_counter(ID)),
    assert(engine(ID, Horsepower, false)).

% Start engine
start_engine(ID) :-
    retract(engine(ID, HP, _)),
    assert(engine(ID, HP, true)),
    format('Engine starting... ~whp engine now running', [HP]).

% Stop engine
stop_engine(ID) :-
    retract(engine(ID, HP, _)),
    assert(engine(ID, HP, false)),
    write('Engine stopped').

% car(CarID, Brand, Model, EngineID) - composition!
:- dynamic car/4.
:- dynamic car_counter/1.
car_counter(0).

% Create new car
new_car(Brand, Model, Horsepower, CarID) :-
    new_engine(Horsepower, EngineID),
    retract(car_counter(Counter)),
    CarID is Counter + 1,
    assert(car_counter(CarID)),
    assert(car(CarID, Brand, Model, EngineID)).

% Start car
start_car(CarID) :-
    car(CarID, Brand, Model, EngineID),
    format('   ~w ~w: ', [Brand, Model]),
    start_engine(EngineID),
    nl.

% Stop car
stop_car(CarID) :-
    car(CarID, Brand, Model, EngineID),
    format('   ~w ~w: ', [Brand, Model]),
    stop_engine(EngineID),
    nl.

% ====================
% 7. Design Pattern: Factory
% ====================

create_animal(dog, Name) :-
    assertz(animal(Name, dog, canine)),
    assertz(dog_breed(Name, 'Mixed')).

create_animal(cat, Name) :-
    assertz(animal(Name, cat, feline)),
    assertz(cat_indoor(Name, true)).

% ====================
% 8. Multiple "Interfaces" with Multiple Predicates
% ====================

% Duck can do multiple things
duck(donald).

% Duck implements Animal
animal(donald, duck, waterfowl).

% Duck-specific abilities
can_fly(donald).
can_swim(donald).
can_quack(donald).

% "Methods" for different abilities
fly(Name) :-
    can_fly(Name),
    format('   ~w is flying through the air~n', [Name]).

swim(Name) :-
    can_swim(Name),
    format('   ~w is swimming in water~n', [Name]).

quack(Name) :-
    can_quack(Name),
    format('   ~w says Quack!~n', [Name]).

% ====================
% 9. "Singleton" with Facts
% ====================

:- dynamic singleton_data/1.
singleton_data([]).

get_singleton_data(Data) :-
    singleton_data(Data).

add_singleton_data(Item) :-
    retract(singleton_data(Data)),
    append(Data, [Item], NewData),
    assert(singleton_data(NewData)).

% ====================
% 10. Point with Operations
% ====================

point(p1, 3, 4).
point(p2, 1, 2).

% "Methods"
add_points(P1, P2, point(X, Y)) :-
    point(P1, X1, Y1),
    point(P2, X2, Y2),
    X is X1 + X2,
    Y is Y1 + Y2.

subtract_points(P1, P2, point(X, Y)) :-
    point(P1, X1, Y1),
    point(P2, X2, Y2),
    X is X1 - X2,
    Y is Y1 - Y2.

distance_from_origin(P, Dist) :-
    point(P, X, Y),
    Dist is sqrt(X * X + Y * Y).

% ====================
% Main Demonstration
% ====================

main :-
    writeln('=== Object-Oriented Programming in Prolog ==='),
    nl,

    % 1. Basic "class"
    writeln('1. Objects as Facts:'),
    introduce(alice),
    (is_adult(alice) -> writeln('   Alice is an adult') ; writeln('   Alice is not an adult')),

    % 2. "Inheritance" and polymorphism
    nl,
    writeln('2. "Inheritance" and Polymorphism:'),
    speak(buddy, _),
    speak(whiskers, _),
    speak(max, _),
    fetch(buddy),
    scratch(whiskers),

    % 3. Shapes
    nl,
    writeln('3. Shapes (Polymorphic Methods):'),
    describe_shape(circle1),
    describe_shape(rect1),
    describe_shape(tri1),

    % 4. Encapsulation (Bank Account)
    nl,
    writeln('4. Encapsulation (Bank Account):'),
    new_account('ACC001', 1000, AccID),
    get_balance(AccID, InitBal),
    format('   Initial balance: $~2f~n', [InitBal]),
    deposit(AccID, 500),
    get_balance(AccID, Bal1),
    format('   After deposit: $~2f~n', [Bal1]),
    withdraw(AccID, 200),
    get_balance(AccID, Bal2),
    format('   After withdrawal: $~2f~n', [Bal2]),
    get_transactions(AccID, Trans),
    format('   Transactions: ~w~n', [Trans]),

    % 5. Temperature
    nl,
    writeln('5. Temperature Conversion:'),
    temperature(celsius, 0, C1),
    to_fahrenheit(C1, F1),
    format('   0°C = ~1f°F~n', [F1]),
    temperature(fahrenheit, 32, C2),
    format('   32°F = ~1f°C~n', [C2]),
    temperature(kelvin, 273.15, C3),
    format('   273.15K = ~1f°C~n', [C3]),
    (is_freezing(0) -> writeln('   Is 0°C freezing? true') ; writeln('   Is 0°C freezing? false')),

    % 6. Composition
    nl,
    writeln('6. Composition:'),
    new_car('Toyota', 'Camry', 200, CarID),
    start_car(CarID),
    stop_car(CarID),

    % 7. Multiple "interfaces"
    nl,
    writeln('7. Multiple Abilities (Duck):'),
    quack(donald),
    fly(donald),
    swim(donald),

    % 8. Points
    nl,
    writeln('8. Point Operations:'),
    point(p1, X1, Y1),
    point(p2, X2, Y2),
    format('   p1 = point(~w, ~w)~n', [X1, Y1]),
    format('   p2 = point(~w, ~w)~n', [X2, Y2]),
    add_points(p1, p2, point(X3, Y3)),
    format('   p1 + p2 = point(~w, ~w)~n', [X3, Y3]),
    subtract_points(p1, p2, point(X4, Y4)),
    format('   p1 - p2 = point(~w, ~w)~n', [X4, Y4]),
    distance_from_origin(p1, Dist),
    format('   p1 distance from origin: ~2f~n', [Dist]),

    % 9. Singleton
    nl,
    writeln('9. Singleton Pattern:'),
    add_singleton_data(item1),
    get_singleton_data(Data1),
    format('   Singleton data: ~w~n', [Data1]),

    % 10. Factory
    nl,
    writeln('10. Factory Pattern:'),
    create_animal(dog, rover),
    create_animal(cat, mittens),
    speak(rover, _),
    speak(mittens, _),

    nl,
    writeln('11. Key Takeaway:'),
    writeln('   Prolog models OOP concepts through logic:'),
    writeln('   - Facts represent object state'),
    writeln('   - Rules represent methods/behavior'),
    writeln('   - Pattern matching provides polymorphism'),
    writeln('   - Dynamic predicates allow state changes'),

    % Exit successfully
    halt(0).

% Handle errors gracefully
:- catch(main, E, (writeln(E), halt(1))).

File: lessons/07-oop/oop.rs
// Lesson 7: Object-Oriented Programming in Rust
//
// Rust doesn't have classes or inheritance in the traditional sense.
// Instead, it uses:
// - Structs for data
// - Impl blocks for methods
// - Traits for shared behavior (like interfaces)
// - Composition over inheritance

use std::f64::consts::PI;

// ====================
// 1. Basic Struct with Methods
// ====================

struct Person {
    name: String,
    age: u32,
}

impl Person {
    // Associated function (like static method) - constructor
    fn new(name: String, age: u32) -> Self {
        Person { name, age }
    }

    // Method (takes &self)
    fn introduce(&self) -> String {
        format!("Hi, I'm {}, {} years old", self.name, self.age)
    }

    // Mutable method (takes &mut self)
    fn have_birthday(&mut self) {
        self.age += 1;
    }

    fn get_age(&self) -> u32 {
        self.age
    }
}

// ====================
// 2. Traits (like interfaces)
// ====================

trait Animal {
    fn speak(&self) -> String;
    fn sleep(&self) -> String {
        // Default implementation
        "Zzz...".to_string()
    }
}

struct Dog {
    name: String,
    breed: String,
}

impl Dog {
    fn new(name: String, breed: String) -> Self {
        Dog { name, breed }
    }

    fn fetch(&self) -> String {
        format!("{} is fetching the ball!", self.name)
    }
}

impl Animal for Dog {
    fn speak(&self) -> String {
        format!("{} says Woof!", self.name)
    }

    fn sleep(&self) -> String {
        format!("{} is sleeping... Zzz", self.name)
    }
}

struct Cat {
    name: String,
    indoor: bool,
}

impl Cat {
    fn new(name: String, indoor: bool) -> Self {
        Cat { name, indoor }
    }

    fn scratch(&self) -> String {
        format!("{} is scratching the furniture!", self.name)
    }
}

impl Animal for Cat {
    fn speak(&self) -> String {
        format!("{} says Meow!", self.name)
    }

    fn sleep(&self) -> String {
        format!("{} is sleeping... Zzz", self.name)
    }
}

// ====================
// 3. Trait Objects for Polymorphism
// ====================

fn make_animal_speak(animal: &dyn Animal) {
    println!("   {}", animal.speak());
}

// ====================
// 4. Abstract-like Behavior with Traits
// ====================

trait Shape {
    fn area(&self) -> f64;
    fn perimeter(&self) -> f64;

    fn describe(&self) -> String {
        format!(
            "{}: area={:.2}, perimeter={:.2}",
            self.name(),
            self.area(),
            self.perimeter()
        )
    }

    fn name(&self) -> &str;
}

struct Circle {
    radius: f64,
}

impl Circle {
    fn new(radius: f64) -> Self {
        Circle { radius }
    }
}

impl Shape for Circle {
    fn area(&self) -> f64 {
        PI * self.radius * self.radius
    }

    fn perimeter(&self) -> f64 {
        2.0 * PI * self.radius
    }

    fn name(&self) -> &str {
        "Circle"
    }
}

struct Rectangle {
    width: f64,
    height: f64,
}

impl Rectangle {
    fn new(width: f64, height: f64) -> Self {
        Rectangle { width, height }
    }
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }

    fn perimeter(&self) -> f64 {
        2.0 * (self.width + self.height)
    }

    fn name(&self) -> &str {
        "Rectangle"
    }
}

struct Triangle {
    side_a: f64,
    side_b: f64,
    side_c: f64,
}

impl Triangle {
    fn new(side_a: f64, side_b: f64, side_c: f64) -> Self {
        Triangle {
            side_a,
            side_b,
            side_c,
        }
    }
}

impl Shape for Triangle {
    fn area(&self) -> f64 {
        // Heron's formula
        let s = self.perimeter() / 2.0;
        (s * (s - self.side_a) * (s - self.side_b) * (s - self.side_c)).sqrt()
    }

    fn perimeter(&self) -> f64 {
        self.side_a + self.side_b + self.side_c
    }

    fn name(&self) -> &str {
        "Triangle"
    }
}

// ====================
// 5. Encapsulation (pub vs private)
// ====================

mod bank {
    pub struct BankAccount {
        account_number: String,
        balance: f64,          // Private by default!
        transactions: Vec<String>,
    }

    impl BankAccount {
        pub fn new(account_number: String, initial_balance: f64) -> Self {
            BankAccount {
                account_number,
                balance: initial_balance,
                transactions: Vec::new(),
            }
        }

        pub fn deposit(&mut self, amount: f64) -> bool {
            if amount > 0.0 {
                self.balance += amount;
                self.transactions
                    .push(format!("Deposit: +${:.2}", amount));
                true
            } else {
                false
            }
        }

        pub fn withdraw(&mut self, amount: f64) -> bool {
            if amount > 0.0 && amount <= self.balance {
                self.balance -= amount;
                self.transactions
                    .push(format!("Withdrawal: -${:.2}", amount));
                true
            } else {
                false
            }
        }

        pub fn get_balance(&self) -> f64 {
            self.balance
        }

        pub fn get_transaction_history(&self) -> Vec<String> {
            self.transactions.clone()
        }
    }
}

use bank::BankAccount;

// ====================
// 6. Associated Functions (like static methods)
// ====================

struct Temperature {
    celsius: f64,
}

impl Temperature {
    // Associated function (constructor)
    fn new(celsius: f64) -> Self {
        Temperature { celsius }
    }

    // Factory methods
    fn from_fahrenheit(fahrenheit: f64) -> Self {
        let celsius = (fahrenheit - 32.0) * 5.0 / 9.0;
        Temperature { celsius }
    }

    fn from_kelvin(kelvin: f64) -> Self {
        let celsius = kelvin - 273.15;
        Temperature { celsius }
    }

    fn is_freezing(celsius: f64) -> bool {
        celsius <= 0.0
    }

    fn to_fahrenheit(&self) -> f64 {
        (self.celsius * 9.0 / 5.0) + 32.0
    }

    fn to_kelvin(&self) -> f64 {
        self.celsius + 273.15
    }
}

// ====================
// 7. Composition
// ====================

struct Engine {
    horsepower: u32,
    running: bool,
}

impl Engine {
    fn new(horsepower: u32) -> Self {
        Engine {
            horsepower,
            running: false,
        }
    }

    fn start(&mut self) -> String {
        self.running = true;
        format!("Engine starting... {}hp engine now running", self.horsepower)
    }

    fn stop(&mut self) -> String {
        self.running = false;
        "Engine stopped".to_string()
    }
}

struct Car {
    brand: String,
    model: String,
    engine: Engine, // Composition!
}

impl Car {
    fn new(brand: String, model: String, horsepower: u32) -> Self {
        Car {
            brand,
            model,
            engine: Engine::new(horsepower),
        }
    }

    fn start(&mut self) -> String {
        format!("{} {}: {}", self.brand, self.model, self.engine.start())
    }

    fn stop(&mut self) -> String {
        format!("{} {}: {}", self.brand, self.model, self.engine.stop())
    }
}

// ====================
// 8. Multiple Traits (like multiple interfaces)
// ====================

trait Flyable {
    fn fly(&self) -> String;
}

trait Swimmable {
    fn swim(&self) -> String;
}

struct Duck {
    name: String,
}

impl Duck {
    fn new(name: String) -> Self {
        Duck { name }
    }

    fn quack(&self) -> String {
        "Quack!".to_string()
    }
}

impl Animal for Duck {
    fn speak(&self) -> String {
        format!("{} says Quack!", self.name)
    }
}

impl Flyable for Duck {
    fn fly(&self) -> String {
        format!("{} is flying through the air", self.name)
    }
}

impl Swimmable for Duck {
    fn swim(&self) -> String {
        format!("{} is swimming in water", self.name)
    }
}

// ====================
// 9. Operator Overloading with Traits
// ====================

use std::ops::{Add, Sub};

#[derive(Debug, Clone, Copy)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    fn distance_from_origin(&self) -> f64 {
        (self.x * self.x + self.y * self.y).sqrt()
    }
}

impl Add for Point {
    type Output = Point;

    fn add(self, other: Point) -> Point {
        Point::new(self.x + other.x, self.y + other.y)
    }
}

impl Sub for Point {
    type Output = Point;

    fn sub(self, other: Point) -> Point {
        Point::new(self.x - other.x, self.y - other.y)
    }
}

// ====================
// 10. Generic Structs and Traits
// ====================

struct Container<T> {
    value: T,
}

impl<T> Container<T> {
    fn new(value: T) -> Self {
        Container { value }
    }

    fn get(&self) -> &T {
        &self.value
    }
}

// ====================
// 11. Design Pattern: Builder
// ====================
// (Rust doesn't use Singleton much due to ownership)

struct PersonBuilder {
    name: Option<String>,
    age: Option<u32>,
}

impl PersonBuilder {
    fn new() -> Self {
        PersonBuilder {
            name: None,
            age: None,
        }
    }

    fn name(mut self, name: String) -> Self {
        self.name = Some(name);
        self
    }

    fn age(mut self, age: u32) -> Self {
        self.age = Some(age);
        self
    }

    fn build(self) -> Result<Person, &'static str> {
        Ok(Person {
            name: self.name.ok_or("Name is required")?,
            age: self.age.ok_or("Age is required")?,
        })
    }
}

// ====================
// Tests and Examples
// ====================

fn main() {
    println!("=== Object-Oriented Programming in Rust ===\n");

    // 1. Basic struct
    println!("1. Basic Struct:");
    let mut alice = Person::new("Alice".to_string(), 30);
    println!("   {}", alice.introduce());
    alice.have_birthday();
    println!("   After birthday: age = {}", alice.get_age());

    // 2. Traits and Polymorphism
    println!("\n2. Traits and Polymorphism:");
    let buddy = Dog::new("Buddy".to_string(), "Golden Retriever".to_string());
    let whiskers = Cat::new("Whiskers".to_string(), true);
    let max = Dog::new("Max".to_string(), "German Shepherd".to_string());

    println!("   {}", buddy.speak());
    println!("   {}", whiskers.speak());
    println!("   {}", max.speak());

    println!("   {}", buddy.fetch());
    println!("   {}", whiskers.scratch());

    // Trait objects for runtime polymorphism
    println!("\n   Using trait objects:");
    let animals: Vec<&dyn Animal> = vec![&buddy, &whiskers, &max];
    for animal in animals {
        make_animal_speak(animal);
    }

    // 3. Shapes
    println!("\n3. Shapes (Trait-based polymorphism):");
    let circle = Circle::new(5.0);
    let rectangle = Rectangle::new(4.0, 6.0);
    let triangle = Triangle::new(3.0, 4.0, 5.0);

    let shapes: Vec<&dyn Shape> = vec![&circle, &rectangle, &triangle];
    for shape in shapes {
        println!("   {}", shape.describe());
    }

    // 4. Encapsulation (Bank Account)
    println!("\n4. Encapsulation (Bank Account):");
    let mut account = BankAccount::new("ACC001".to_string(), 1000.0);
    println!("   Initial balance: ${:.2}", account.get_balance());
    account.deposit(500.0);
    println!("   After deposit: ${:.2}", account.get_balance());
    account.withdraw(200.0);
    println!("   After withdrawal: ${:.2}", account.get_balance());
    println!("   Transactions: {:?}", account.get_transaction_history());

    // 5. Associated functions
    println!("\n5. Associated Functions:");
    let temp1 = Temperature::new(0.0);
    let temp2 = Temperature::from_fahrenheit(32.0);
    let temp3 = Temperature::from_kelvin(273.15);

    println!("   0°C = {:.1}°F", temp1.to_fahrenheit());
    println!("   32°F = {:.1}°C", temp2.celsius);
    println!("   273.15K = {:.1}°C", temp3.celsius);
    println!("   Is 0°C freezing? {}", Temperature::is_freezing(0.0));

    // 6. Composition
    println!("\n6. Composition:");
    let mut car = Car::new("Toyota".to_string(), "Camry".to_string(), 200);
    println!("   {}", car.start());
    println!("   {}", car.stop());

    // 7. Multiple traits
    println!("\n7. Multiple Traits:");
    let duck = Duck::new("Donald".to_string());
    println!("   {}", duck.quack());
    println!("   {}", duck.fly());
    println!("   {}", duck.swim());

    // 8. Operator overloading
    println!("\n8. Operator Overloading:");
    let p1 = Point::new(3.0, 4.0);
    let p2 = Point::new(1.0, 2.0);
    let p3 = p1 + p2;
    let p4 = p1 - p2;
    println!("   p1 = {:?}", p1);
    println!("   p2 = {:?}", p2);
    println!("   p1 + p2 = {:?}", p3);
    println!("   p1 - p2 = {:?}", p4);
    println!(
        "   p1 distance from origin: {:.2}",
        p1.distance_from_origin()
    );

    // 9. Generics
    println!("\n9. Generics:");
    let int_container = Container::new(42);
    let string_container = Container::new("Hello".to_string());
    println!("   Integer container: {}", int_container.get());
    println!("   String container: {}", string_container.get());

    // 10. Builder pattern
    println!("\n10. Builder Pattern:");
    let bob = PersonBuilder::new()
        .name("Bob".to_string())
        .age(25)
        .build()
        .unwrap();
    println!("   {}", bob.introduce());
}
